### PR TITLE
feat(contrib/coreos): move static user-data outside

### PR DESCRIPTION
**Pull request by aespinosa**
Significantly decreases user-data content to only the ones needed for
configuration.

Proposed idea for #3751

### CODE REVIEW

**Comment by deis-admin** [comment]
Thanks for the contribution! Please ensure your commits follow our [style guide](http://docs.deis.io/en/latest/contributing/standards/#commit-style-guide). This code will be tested once a Deis maintainer reviews it.

Your Pull Request can earn you credit at DigitalOcean! See http://docs.deis.io/en/latest/contributing/hacking/#digitalocean-credit for details.

File: contrib/coreos/user-data.example [hunk 1]
```diff
@@ -17,266 +17,13 @@ coreos:
...
-      fi
-  - path: /opt/bin/scheduler-policy.json
-    content: |
-      {
-          "kind": "Policy",
-          "apiVersion": "v1",
-          "predicates": [{"name": "PodFitsPorts"},{"name": "PodFitsResources"},{"name": "NoDiskConflict"},{"name": "MatchNodeSelector"},{"name": "HostName"}],
-          "priorities": [{"name": "LeastRequestedPriority","weight": 1},{"name": "BalancedResourceAllocation","weight": 1},{"name": "ServiceSpreadingPriority","weight": 2},{"name": "EqualPriority","weight": 1}]
-      }
+        ExecStart=/usr/bin/coreos-cloudinit --from-url=https://raw.githubusercontent.com/aespinosa/deis/move-static-cloud-config/contrib/coreos/cloud-config
```
Comment (aespinosa) [inline contrib/coreos/user-data.example]: I just set this to my branch to pass the tests. This should be something canonical and managed synced with releases.

File: contrib/coreos/cloud-config [hunk 2]
```diff
@@ -0,0 +1,266 @@
+#cloud-config
```
Comment (bacongobbler) [inline contrib/coreos/cloud-config]: Would you mind removing this file or moving it somewhere else if this is meant to be the static cloud config to be uploaded to S3?

**Comment by krancour** [comment]
Jenkins, add to whitelist.

File: contrib/coreos/cloud-config (hunk 2 above)
Comment (krancour) [inline contrib/coreos/cloud-config]: @bacongobbler isn't it `contrib/coreos/user-data` that gets written by `make discovery-url`?

**Comment by krancour** [comment]
I like this change, mainly because it reduces the size of the artifacts that get submitted to each cloud provider during provisioning-- e.g. smaller CloudFormation templates.  That means we can put _more_ stuff in there (to make clusters more robust) without fearing the size restrictions too much.

All that being said...

I think there are some gotchas that this PR does not account for.  Take the AWS provisioning process as an example.  That process actually merges `contrib/coreos/user-data` with additional AWS-specific user-data.  You can see that [here](https://github.com/deis/deis/blob/master/contrib/aws/gen-json.py#L15-L106).  Will that continue to work?  Or will it break?

My one other concern would be with respect to workflow.  Does this make it much harder for people to customize their user-data if they have good reason to do so?

I still like this.  I just want to make sure everything is accounted for.

File: contrib/coreos/cloud-config (hunk 2 above)
Comment (bacongobbler) [inline contrib/coreos/cloud-config]: Yes, which is why this file should be moved elsewhere or else `make discovery-url` will overwrite it. This file is in gitignore: https://github.com/deis/deis/blob/master/.gitignore#L68-L69

Comment (carmstrong) [inline contrib/coreos/cloud-config]: This file is `cloud-config`, not `user-data`.

Comment (bacongobbler) [inline contrib/coreos/cloud-config]: Wow, big oversight by me. Need more :coffee:! Sorry <3

**Comment by carmstrong** [comment]
Agreed with @krancour - I love this in principle, but I'm thinking about the dependency on GitHub for a machine to bootstrap properly. What if GitHub is down?

**Comment by aespinosa** [comment]
Hi @carmstrong 

Yes I agree that the release engineering of where this artifact should be needs some thinking. Right now user-data as-is depends on some github/ google-located artifacts. Is there a place where we consider "most reliable"? Or are we headed to the bosh/cf-style assumption of everything should be vendored on the machine doing the provisioning?

@krancour , right I haven't taken the IaaS-specific user-data being injected. That being said, I agree that managing YML-ish manifests is a hard problem. Several related tooling have come and go. Here's a few:

* https://github.com/mattray/spiceweasel
* https://github.com/cloudfoundry-incubator/spiff
* https://github.com/geofffranks/spruce

They're very specific to their own domains but I don't seem to find coreos folks thinking this is a problem with cloudinit.

**Comment by carmstrong** [comment]
@aespinosa Good points. I think @krancour's point about provider-specific cloud-config is valid. I'm wondering if we should build provider-specific cloud-config files whenever we cut a Deis release, and push those somewhere public like S3. Added complexity at release time, but may be better for users.

**Comment by krancour** [comment]
@carmstrong, that sounds like a maintenance burden, but I agree that it simplifies an awful lot of things.

**Comment by bacongobbler** [comment]
closing as something we won't be able to maintain during the LTS release as mentioned in #3751. Thank you for the PR though :)

**Pull request closed without merging**